conjure-core/src/main/java/com/palantir/conjure/defs/validator/EndpointDefinitionValidator.java

```diff
@@ -100,17 +101,32 @@ public void validate(EndpointDefinition definition, DealiasingTypeVisitor dealia
                     .stream()
                     .filter(arg -> !arg.getParamType().accept(ParameterTypeVisitor.IS_BODY))
                     .forEach(arg -> {
-                        boolean isValid = dealiasingTypeVisitor.dealias(arg.getType())
-                                .fold(
-                                        typeDefinition -> true,
-                                        type -> !type.accept(TypeVisitor.IS_BINARY)
-                                                && !type.accept(TypeVisitor.IS_ANY)
-                                );
                         Preconditions.checkArgument(
-                                isValid, "Non body parameters cannot be of the 'binary' type: '%s' is not allowed",
+                                validateType(arg.getType(), dealiasingTypeVisitor),
+                                "Non body parameters cannot be of the 'binary' type: '%s' is not allowed",
                                 arg.getArgName());
                     });
         }
+
+        private static boolean validateType(Type input, DealiasingTypeVisitor dealiasingTypeVisitor) {
+            Optional<Type> dealiased = dealiasingTypeVisitor.dealias(input)
+                    .fold(typeDefinition -> Optional.empty(), Optional::of);
+            // typeDef isn't binary
+            if (!dealiased.isPresent()) {
+                return true;
+            }
+            Type type = dealiased.get();
+            if (input.accept(TypeVisitor.IS_OPTIONAL)) {
+                return validateType(input.accept(TypeVisitor.OPTIONAL).getItemType(), dealiasingTypeVisitor);
+            }
+            if (input.accept(TypeVisitor.IS_LIST)) {
+                return validateType(input.accept(TypeVisitor.LIST).getItemType(), dealiasingTypeVisitor);
+            }
+            if (input.accept(TypeVisitor.IS_SET)) {
+                return validateType(input.accept(TypeVisitor.SET).getItemType(), dealiasingTypeVisitor);
+            }
+            return !type.accept(TypeVisitor.IS_BINARY) && !type.accept(TypeVisitor.IS_ANY);
```

Comment:
Not sure if this should be handled in a separate PR, but for query and path parameters, the check should also reject containers that contain `bearertoken`.